stixcore/processing/level0.py

```diff
@@ -0,0 +1,365 @@
+        if spid:
+            path = basepath
+            path.mkdir(exist_ok=True, parents=True)
+            # try:
+            process_packets(data, spid, product, path, status=type, overwrite=overwrite,
+                            use_name=use_name)
+            # except Exception as e:
+            #     logger.error('error %d, %s', spid, product, exc_info=True)
+
+
+def fits_to_packets(file):
+    logger.info(f'Processing fits file {file}')
+    control = Table.read(str(file), hdu=1)
+    data = Table.read(str(file), hdu=2)
+
+    packets = [Packet(unhexlify(d)) for d in data['data']]
+    if np.abs([((len(data['data'][i])//2) - (control['data_len'][i]+7))
+               for i in range(len(data))]).sum() > 0:
+        raise ValueError('Packet size and expected length do not match')
+    # packets = list(chain.from_iterable(packets))
+    # Filter keeping only TM packets
+    # packets = list(filter(lambda x: x['header']['TMTC'] == 'TM', packets))
+    # Packet ordering is not guaranteed so sort by coarse time then seq count
+    # packets.sort(key=lambda x: (x['header']['coarse_time'], x['header']['seq_count']))
+    return packets
+
+
+class Level0:
+    def __init__(self, source_dir, output_dir, use_name=False):
+        self.source_dir = Path(source_dir)
+        self.output_dir = Path(output_dir)
+        self.level0_files = sorted(list(self.source_dir.rglob('*.fits')),  key=os.path.getctime)
+        self.level0_files = self.level0_files[-17:]
+        self.use_name = use_name
+
+    def process_fits_files(self):
+        cur_packets = fits_to_packets(self.level0_files[0])
+        cur_complete, cur_incomplete = get_products(cur_packets)
+        spid = set(list(cur_complete.keys()) + list(cur_incomplete.keys()))
+        if len(spid) > 1:
+            raise ValueError()
+        spid = spid.pop()
+        if len(self.level0_files) > 1:
+            for file in self.level0_files[1:]:
+                next_packets = fits_to_packets(file)
+                next_complete, next_incomplete = get_products(next_packets)
+
+                if len(cur_incomplete[spid]) > 0 and len(next_incomplete[spid]) > 0:
+                    incomplete_combined, incomplete_remaining\
+                        = get_products(cur_incomplete[spid][0] + next_incomplete[spid][0])
+
+                    if len(incomplete_combined) > 0:
+                        complete = cur_complete.copy()
+                        complete[spid] = [complete[spid][0] + incomplete_combined[spid][0]]
+                        incomplete = {}
+                        next_incomplete = incomplete_remaining
+                    else:
+                        complete = cur_complete
+                        incomplete = cur_incomplete
+                else:
+                    complete = cur_complete
+                    incomplete = cur_incomplete
+
+                cur_complete = next_complete
+                cur_incomplete = next_incomplete
```

Comment:
This is all trying to make sure complete packets are that are separated over day are combined